3scaleAdapter/pkg/threescale/proxy_conf.go

```diff
@@ -0,0 +1,261 @@
+	if pc.flushWorkerRunning {
+		return errors.New("worker has already been started")
+	}
+
+	pc.stopFlushWorker = make(chan bool)
+	go pc.flushCache(pc.stopFlushWorker)
+	pc.flushWorkerRunning = true
+	return nil
+}
+
+// StopFlushWorker stops a background process started by StartFlushWorker if it has been started
+// Returns an error if the background task is not running
+func (pc *ProxyConfigCache) StopFlushWorker() error {
+	if pc.stopFlushWorker == nil || !pc.flushWorkerRunning {
+		return errors.New("worker has not been started")
+	}
+
+	pc.stopFlushWorker <- true
+	close(pc.stopFlushWorker)
+	pc.flushWorkerRunning = false
+	return nil
+}
+
+// StartRefreshWorker starts a background process that periodically carries out the
+// functionality provided by RefreshCache
+func (pc *ProxyConfigCache) StartRefreshWorker() error {
+	if pc.refreshWorkerRunning {
+		return errors.New("worker has already been started")
+	}
+
+	pc.stopRefreshWorker = make(chan bool)
+	go pc.refreshCache(pc.stopRefreshWorker)
+	pc.refreshWorkerRunning = true
+	return nil
+}
+
+// StopRefreshWorker stops a background process started by StartRefreshWorker if it has been started
+// Returns an error if the background task is not running
+func (pc *ProxyConfigCache) StopRefreshWorker() error {
+	if pc.stopRefreshWorker == nil || !pc.refreshWorkerRunning {
+		return errors.New("worker has not been started")
+	}
+
+	pc.stopRefreshWorker <- true
+	close(pc.stopRefreshWorker)
+	pc.refreshWorkerRunning = false
+	return nil
+}
+
+func (pc *ProxyConfigCache) get(cfg *config.Params, c *sysC.ThreeScaleClient) (sysC.ProxyConfigElement, error) {
+	var conf sysC.ProxyConfigElement
+	var err error
+
+	cacheKey := pc.getCacheKeyFromCfg(cfg)
+
+	pc.mutex.RLock()
+	e, ok := pc.cache[cacheKey]
+	pc.mutex.RUnlock()
+
+	if !ok {
+		conf, err = getFromRemote(cfg, c)
+		if err == nil {
+			replayWith := cacheRefreshStore{cfg, c}
+			go pc.set(cacheKey, conf, replayWith)
+		}
+	} else {
+		log.Debugf("proxy config fetched from cache for service id %s", cfg.ServiceId)
+		conf = e.element
+	}
+
+	return conf, err
+}
+
+func (pc *ProxyConfigCache) set(cacheKey string, e sysC.ProxyConfigElement, replayWith cacheRefreshStore) {
+	expiresAt := time.Now().Add(pc.ttl)
+	pc.mutex.Lock()
+	defer pc.mutex.Unlock()
+	if len(pc.cache) < pc.limit {
+		pc.cache[cacheKey] = proxyStore{
+			expiresAt:  expiresAt,
+			element:    e,
+			replayWith: replayWith,
+		}
+	}
+}
+
+func (pc *ProxyConfigCache) delete(key ...string) {
+	pc.mutex.Lock()
+	for _, k := range key {
+		delete(pc.cache, k)
+	}
+	pc.mutex.Unlock()
+}
+
+func (pc *ProxyConfigCache) getCacheKeyFromCfg(cfg *config.Params) string {
+	return fmt.Sprintf(cacheKey, cfg.SystemUrl, cfg.ServiceId)
+}
+
+func (pc *ProxyConfigCache) flushCache(exitC chan bool) {
+	for {
+
+		select {
+		case stop := <-exitC:
+			if stop {
+				log.Debugf("stopping cache flushing worker")
+				return
+			}
+		default:
+			pc.FlushCache()
+			<-time.After(pc.ttl)
+		}
+
```

Comment:
some extra blank lines here in this loop, is this idiomatic in Go?

Reply:
No, will fix